aya/src/maps/mod.rs

```diff
@@ -281,155 +281,122 @@ impl Map {
     }
 }
 
+// Implements TryFrom<Map> for different map implementations. Different map implementations can be
+// constructed from different variants of the map enum. Also, the implementation may have type
+// parameters (which we assume all have the bound `Pod` and nothing else).
 macro_rules! impl_try_from_map {
-    ($($tx:ident from Map::$ty:ident),+ $(,)?) => {
-        $(
-            impl<'a> TryFrom<&'a Map> for $tx<&'a MapData> {
-                type Error = MapError;
-
-                fn try_from(map: &'a Map) -> Result<$tx<&'a MapData>, MapError> {
-                    match map {
-                        Map::$ty(m) => {
-                            $tx::new(m)
-                        },
-                        _ => Err(MapError::InvalidMapType{ map_type: map.map_type()}),
-                    }
-                }
+    // Wrap patterns without type parameters.
+    (
+        impl {
+            $($ty:ident $(from $variant:ident)?),+
+            $(,)?
+        }
+    ) => {
+        impl_try_from_map! {
+            impl <()> {
+                $($ty $(from $variant)?),+
             }
-
-            impl<'a,> TryFrom<&'a mut Map> for $tx<&'a mut MapData> {
-                type Error = MapError;
-
-                fn try_from(map: &'a mut Map) -> Result<$tx<&'a mut MapData>, MapError> {
-                    match map {
-                        Map::$ty(m) => {
-                            $tx::new(m)
-                        },
-                        _ => Err(MapError::InvalidMapType{ map_type: map.map_type()}),
-                    }
-                }
+        }
+    };
+    // Wrap the type parameters in parentheses so that they can be treated as a single TokenTree
+    // (tt) because you cannot compose repetitions but you can expand a scalar in a repetition.
```

Comment:
i think this comment could be clearer; the point is that we're not interested in repetition, we just want to paste all the type parameters in a few places.

I think everything here will be simpler if you push the parentheses out into the macro invocation and include empty parentheses in the case of no type parameters.